### ISSUE TITLE
Output ledger format journal

### ISSUE THREAD

**Issue by colindean** [hrj/abandon#150]
I would expect these to output the same data but in different formats:

```
exports += {
  type = journal
  format = xml
  outFiles = [reports/transactions.xml]
}

exports += {
  type = journal
  format = ledger
  outFiles = [reports/transactions.ledger]
}
```

The first outputs _all_ transactions in XML format. Great for analysis with XML tools, bad for using other ledger-like tools for analysis.

The second outputs _only_ a single, closing transaction. Great for creating a closing transaction, bad for using other ledger-like tools for analysis.

It'd really like for some report to spit out all of the transactions in ledger format.

**Comment by colindean** [comment]
I made an attempt at this a while ago, simply transforming the XML in Scala. 

```scala

//!/usr/bin/env scala

import scala.xml.{XML, Node}
import scala.language.postfixOps

val xml = XML.loadFile("reports/transactions.xml")

val txnsXml = xml \ "journal" \ "transactions" \ "txn"

case class Post(account: String, delta: Int) {
  def ledgerString: String = s"$account\t\t${deltaAsDecimal}"
  def deltaAsDecimal: String = "%6.2f".format(delta / 100.00)
}
object Post {
  def toCents(amount: Double): Int = (amount * 100.00).toInt
  def fromXml(node: Node): Post = {
    Post(
      delta   = toCents((node \@ "delta").toDouble),
      account = (node \@ "name")
    )
  }
}

case class Transaction(payee: String, date: String, posts: Seq[Post]) {
  def ledgerString() = {
    s"""|$date  $payee
        |${posts.map(_.ledgerString).map{p => s"  $p"}.mkString("\n")}
    """.stripMargin
  }
}
object Transaction {
  def fromXml(node: Node): Transaction = {
    val posts: Seq[Post] = (node \ "post").map(Post.fromXml)
    Transaction(
      posts = posts,
      date  = (node \@ "date"),
      payee = (node \ "payee") text
    )
  }
}

val txns = txnsXml.map(Transaction.fromXml)

val result = txns.filter { _.posts.exists { _.account.contains(args.mkString) } }

result.foreach { txn => println(txn.ledgerString)}

```

However, I think _I_ might have introduced a rounding error somewhere because the output creates a lot of unbalanced transactions when I try to run it through ledger. Everything that is unbalanced is off by `0.01` or `0.02`.

**Comment by colindean** [comment]
Why do I want this? I want to be able to produce a statement for certain accounts that aligns with the associated institution's statement. This will ease cross-checking and verification. 

I'm presently off my almost $1,100 on one account despite having double checked the statements thrice 😢

**Comment by hrj** [comment]
@colindean To address your immediate worry, isn't this doing the rounding:

```
def toCents(amount: Double): Int = (amount * 100.00).toInt
```

Is it possible to simply use the amount without converting to rounded off cents?

**Comment by colindean** [comment]
Yeah. I was trying to do the right thing and keep everything in cents as integers but clearly I didn't do it right.

I had a flash of insight and did up this XSLT in a few minutes, run with `xsltproc abandon2ledger.xsl reports/transactions.xml`

`abandon2ledger.xsl`:

```xsl
<?xml version="1.0"?>
<xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform">

  <xsl:output method="text" encoding="utf-8"/>

  <xsl:variable name="newline"><xsl:text>&#10;</xsl:text></xsl:variable>
  <xsl:variable name="tab"><xsl:text>&#09;</xsl:text></xsl:variable>

  <xsl:template match="@*|node()">
    <xsl:apply-templates select="@*|node()"/>
  </xsl:template>

  <xsl:template match="txn">
     <xsl:value-of select="concat(@date, ' ', payee, $newline)"/>
     <xsl:for-each select="post">
       <xsl:value-of select="concat($tab, @name, $tab, $tab, @delta, $newline)"/>
     </xsl:for-each>
     <xsl:value-of select="$newline"/>
  </xsl:template>
</xsl:stylesheet>

```

Everything balances now and it's faster than the Scala version!

**Comment by hrj** [comment]
> I had a flash of insight and did up this XSLT in a few minutes

That was slick! Thanks for sharing.

I have a Apache FOP based reporting too that takes the XML and spews out a PDF report. Will open-source it the next time I get a chance.

**Comment by colindean** [comment]
I'd definitely like to see that.

**Comment by hrj** [comment]
> I'd definitely like to see that.

Managed to squeeze some time today for publishing it:
https://github.com/hrj/abandon-reports

:tada:

**Comment by hrj** [comment]
A fix for this is now in PR #211. It's working fine and tests are passing too.

When the PR is merged, there will be two export types (`journal` and `balance`) that determine the content of the exported data, and two format types (`ledger` and `xml`) which only determine the file format.

The combination of `journal` and `ledger` is not implemented yet. Also closures are not supported when file format is `xml`, but support can be added later. If there's need for it, do let me know.

Since this will be a backwards incompatible change, I will bump up the major version in the next release.

Any existing config files that exported with `type:journal & format:ledger` will now have to change it to `type:balance & format:ledger`.
